source/extensions/filters/listener/connect_handler/connect_handler.cc

```diff
@@ -0,0 +1,183 @@
+            cb_->continueFilterChain(false);
+            break;
+          case ParseState::Done:
+            cb_->socket().ioHandle().resetFileEvents();
+            // Do not skip following listener filters.
+            cb_->continueFilterChain(true);
+            break;
+          case ParseState::Continue:
+            // do nothing but wait for the next event
+            break;
+          }
+        },
+        Event::PlatformDefaultTriggerType, Event::FileReadyType::Read);
+    return Network::FilterStatus::StopIteration;
+  }
+  ENVOY_BUG(false, "Unexpected ParseState value for connect handler");
+}
+
+ParseState Filter::onRead() {
+  auto result = cb_->socket().ioHandle().recv(buf_, Config::MAX_INSPECT_SIZE, MSG_PEEK);
+  ENVOY_LOG(trace, "connnect handler: recv: {}", result.return_value_);
+  if (!result.ok()) {
+    if (result.err_->getErrorCode() == Api::IoError::IoErrorCode::Again) {
+      return ParseState::Continue;
+    }
+    config_->stats().read_error_.inc();
+    return ParseState::Error;
+  }
+
+  // Remote closed
+  if (result.return_value_ == 0) {
+    return ParseState::Error;
+  }
+  return parseConnect(absl::string_view(reinterpret_cast<const char*>(buf_), result.return_value_));
+}
+
+ParseState Filter::parseConnect(absl::string_view data) {
+  const size_t len = std::min(data.length(), HTTP1_CONNECT_PREFACE.length());
+  if (HTTP1_CONNECT_PREFACE.compare(0, len, data, 0, len) == 0) {
+    if (data.length() < HTTP1_CONNECT_PREFACE.length()) {
+      return ParseState::Continue;
+    } else {
+      absl::string_view new_data = data.substr(parser_.nread);
+      size_t pos = new_data.find("\r\n\r\n");
+      if (pos != absl::string_view::npos) {
+        new_data = new_data.substr(0, pos + 4);
+        ssize_t rc = http_parser_execute(&parser_, &settings_, new_data.data(), new_data.length());
+        ENVOY_LOG(trace, "connect handler: http_parser parsed {} chars, error code: {}", rc,
+                  HTTP_PARSER_ERRNO(&parser_));
+
+        // Errors in parsing HTTP.
+        if (HTTP_PARSER_ERRNO(&parser_) != HPE_OK && HTTP_PARSER_ERRNO(&parser_) != HPE_PAUSED) {
+          ENVOY_LOG(trace, "connect handler: errors in parsing HTTP");
+          config_->stats().connect_not_found_.inc();
+          return ParseState::Done;
+        }
+        if (parser_.http_major == 1) {
+          if (parser_.http_minor == 0) {
+            protocol_ = Http::Headers::get().ProtocolStrings.Http10String;
+          } else if (parser_.http_minor == 1) {
+            protocol_ = Http::Headers::get().ProtocolStrings.Http11String;
+          }
+        } else {
+          ENVOY_LOG(trace, "connect:handler: unsupported http version: {}.{}", parser_.http_major,
+                    parser_.http_minor);
+          config_->stats().connect_not_found_.inc();
+          return ParseState::Done;
+        }
+
+        std::vector<absl::string_view> fields = absl::StrSplit(data, absl::ByAnyChar(" :"));
```

Comment:
Would it be more efficient to use `StringUtil::cropLeft()` and `StringUtil::cropRight()` for server name extraction?

Reply:
done